keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/NUTS.java

```diff
@@ -0,0 +1,546 @@
+                    logOfMasterPAtPositionBackward,
+                    positionForward,
+                    gradientForward,
+                    momentumForward,
+                    logOfMasterPAtPositionForward,
+                    thetaPrime,
+                    gradientAtThetaPrime,
+                    logOfMasterPAtThetaPrime,
+                    sampleAtThetaPrime,
+                    nPrime,
+                    sPrime
+            );
+        }
+
+    }
+
+    private static double getLogProb(List<Vertex<?>> probabilisticVertices) {
+        double sum = 0.0;
+        for (Vertex<?> vertex : probabilisticVertices) {
+            sum += vertex.logProbAtValue();
+        }
+        return sum;
+    }
+
+    private static boolean withProbability(double probability, Random random) {
+        return random.nextDouble() < probability;
+    }
+
+    private static int isNotUTurning(Map<String, Double> positionForward,
+                                     Map<String, Double> positionBack,
+                                     Map<String, Double> momentumForward,
+                                     Map<String, Double> momentumBack) {
+        double forward = 0.0;
+        double backward = 0.0;
+
+        for (Map.Entry<String, Double> forwardEntry : positionForward.entrySet()) {
+
+            final String id = forwardEntry.getKey();
+            final double forwardMinusBack = forwardEntry.getValue() - positionBack.get(id);
+
+            forward += forwardMinusBack * momentumForward.get(id);
+            backward += forwardMinusBack * momentumBack.get(id);
+        }
+
+        boolean notTurning = forward >= 0.0 && backward >= 0.0;
+
+        return notTurning ? 1 : 0;
+    }
+
+    private static void cachePosition(List<Vertex<Double>> latentVertices, Map<String, Double> position) {
+        for (Vertex<Double> vertex : latentVertices) {
+            position.put(vertex.getId(), vertex.getValue());
+        }
+    }
+
+    private static Map<String, Double> initializeMomentumForEachVertex(List<Vertex<Double>> vertexes,
+                                                                       Map<String, Double> momentums,
+                                                                       Random random) {
+        for (int i = 0; i < vertexes.size(); i++) {
+            Vertex currentVertex = vertexes.get(i);
+            momentums.put(currentVertex.getId(), random.nextGaussian());
+        }
+        return momentums;
+    }
+
+    private static void cache(Map<String, Double> from, Map<String, Double> to) {
+        for (Map.Entry<String, Double> entry : from.entrySet()) {
+            to.put(entry.getKey(), entry.getValue());
+        }
+    }
+
+    /**
+     * function Leapfrog(θ, r, eps)
+     * Set ˜r ← r + (eps/2)∇θL(θ)
+     * Set ˜θ ← θ + r˜
+     * Set ˜r ← r˜ + (eps/2)∇θL(˜θ)
+     * return ˜θ, r˜
+     */
+    private static LeapFrogged leapfrog(final List<Vertex<Double>> latentVertices,
```

Comment:
Is this the same as the hamiltonian `leapfrog`? I know they return different types but is there shared behaviour we could pull out?

Reply:
HMC doesn't need as much information back because the arguments are just mutated. I'd like to avoid tight coupling between the two algorithms until at least the final changes are made to NUTS.